src/selfplay/game.cc

```diff
@@ -91,19 +97,39 @@ SelfPlayGame::SelfPlayGame(PlayerOptions white, PlayerOptions black,
     tree_[1] = std::make_shared<NodeTree>();
     tree_[1]->ResetToPosition(orig_fen_, {});
   }
+  int ply = 0;
+  auto white_prob = white.uci_options->Get<float>(kOpeningStopProbId);
+  auto black_prob = black.uci_options->Get<float>(kOpeningStopProbId);
+  if (white_prob != black_prob && white_prob * black_prob > 0.0f) {
+    throw Exception("Stop probabilities must be both equal or zero!");
+  }
+
   for (Move m : opening.moves) {
+    auto exit_prob_now = tree_[0]->IsBlackToMove() ? black_prob : white_prob;
+    auto exit_prob_next = tree_[0]->IsBlackToMove() ? white_prob : black_prob;
+    int positions = opening.moves.size() - ply + 1;
+    if (exit_prob_now > 0.0f &&
+        Random::Get().GetFloat(1.0f) <
+            std::max(exit_prob_now,
+                     exit_prob_now / (exit_prob_now * ((positions + 1) / 2) +
```

Comment:
Good suggestion, though being concise might be difficult. The idea is that there are 2 possible modes of using this feature: exiting from any point in a PGN, and only exit at white or black moves. At the same time, we assume that the PGNs are selected for training purposes so the early plies are more important than later plies.

This formula ensures that the total exit probability is indeed never favouring the last position over earlier ones (which would happen with a fixed `x%` exit at each ply) by using at least probability `1/N` with `N` being the remaining positions in the book. It looks more complicated because the calculation of the remaining positions depends on whether exits at both sides are allowed, or only one.

Reply:
Added a descriptive comment.